### ISSUE TITLE
please add a convenient means to get joint angles after solving IK

### ISSUE THREAD

**Issue by adamscottv** [EGjoni/Everything-WIll-Be-IK#2]
Currently, an additional rotation calculation step is needed to get joint angles after solving for IK. Since the purpose of IK is to find joint parameters that provide a given end effector position, it is my opinion that an IK library should provide a means to get the joint angle. In this case, something like double bone.getXAngleRadians(), bone.getYAngleRadians(), bone.getZAngleRadians() or a single method like double[] bone.getXYZAnglesRadians() would be nice to have.

**Comment by EGjoni** [comment]
Acknowledged. Will add soon.

**Comment by EGjoni** [comment]
Added a Bone.getXYZAngle() method. Let me know if it works for your purposes.

**Comment by adamscottv** [comment]
Sweet, thank you Eron. I will have some time this weekend to work on the project. I'd like to have some basic IK motions working by Sunday.


BTW, is your name pronounced like Aaron?


-Adam

________________________________

**Comment by adamscottv** [comment]
Had another question I was thinking about. Not sure if this is something your code could easily support or if it would be a lot of work.

In robotics it's nice to be able to us Forward Kinematics (FK) to get xyz positions, especially of the end-effector, given a set of input angles. So it would be kind of the opposite process - call setXYZAngles, then call solveFK, then get vectors as result.

There are some practical applications of this functionality. You can get the actual xyz position vs the target position, in case you didn't actually there. It also allows you to do more types of dynamic adjustments if you know where you are. Another convenience is being able to manually arrange the bones, then get the resulting tip position in order to get IK target positions to use.

I could add an issue request if you think it's something you may consider adding in the future. But I understand if it's not a priority.

-Adam

**Comment by EGjoni** [comment]
Not entirely sure if I understand your question. But as an attempt to
answer it any way:

You can rotate any bone using FK by just calling any one of:

Bone.rotAboutFrameX(rad)
Bone.rotAboutFrameY(rad)
Bone.rotAboutFrameZ(rad)

Or by composing your own arbitrary Rotation using the Apache Commons
Rotation library (included in the .jar) and then calling

Bone.localAxes().rotateTo(Rotation);

You can then just call Bone.getTip() to find where the tip of any Bone is
in global space.

So, basically, if you want to figure out where the toe is after rotating
the femur,
just rotate the femur, then call toe.getTip().

You don't need to call anything to solve FK, because each bone's coordinate
system is defined relative to its parent bone's coordinate system.

**Comment by adamscottv** [comment]
Yes, you did answer it. I just rotate the bone using rotAboutFrame and the tip position is automatically recalculated. I didn't realize that calculation would happen without solving something. That's cool. I'll play around with it after I get some other stuff working.


Thanks,

-Adam

________________________________
